### ISSUE TITLE
perf acceptance criteria - throughput is affected with during ops, when using 50% of max throughput

### ISSUE THREAD

**Issue by fgelcer** [scylladb/scylladb#12212]
as part of the acceptance criteria for the performance tests, we started modifying the tests to attend the requested changed.
one of these requests was to run the test using 50% of the max throughput for that cluster (we have a test that does a series of workload, incrementing it, until reaching the max throughput).

running a test at 50% of max throughput for a cluster of 3 nodes, using `i3en.2xlarge`, we have the following command for mixed workload:
```
"cassandra-stress mixed no-warmup cl=QUORUM duration=800m -schema 'replication(factor=3)' -mode cql3 native -rate 'threads=50 fixed=17000/s' -col 'size=FIXED(128) n=FIXED(8)' -pop 'dist=gauss(1..650000000,325000000,1200000)' "
```

1st of all, the distribution gave us around 99% cache hit ratio (we are changing it, but still) throughput could not stay stable when starting manager repair operation (at 13:09):

![Screenshot from 2022-12-06 10-36-16](https://user-images.githubusercontent.com/32832995/205861503-515292aa-f973-48cd-b467-73c2fb916461.png)

and of course, latency for this run was on hundreds of seconds.

[the real monitor is alive, and can be accessed here](http://3.249.247.170:3000/d/yUm3J-KVz/perf-regression-latency-250gb-with-nemesis-scylla-per-server-metrics-nemesis-master?orgId=1&from=1670242574972&to=1670246371546&var-by=instance&var-cluster=&var-dc=All&var-node=All&var-shard=All&var-sct_tags=InfoEvent)

while after that point, i started using the very same cluster for adjusting the cache hit ratio, the whole cluster is still alive

seems like logs were not collected, but can be found i live nodes

**Comment by fgelcer** [comment]
@roydahan , FYI

**Comment by dorlaor** [comment]
It's not clear whether 100k ops is the max or is 50% of max. It's
relatively high to be the max itself.

**Comment by avikivity** [comment]
Please post a monitor snapshot for me to download.

**Comment by michoecho** [comment]
Besides, there's a population phase some time earlier, and 100k/s seems to be the max throughput, not 50%.

**Comment by roydahan** [comment]
I'll try to make it a bit clearer, because @fgelcer made a mistake here and isn't very clear.
1. @fgelcer confused between the mixed test and the 100% write test - both suffer from the same issue, but the monitor he provided is for write test and the sample command is for mixed test.
2. IIRC the max throughput he got was 200K, so he set the target op/s to 100K.
3. When we set the target ops, we have 4 loaders, so each gets 25K ops target (for mix it's 17K to reach a 68K ops target).

**Comment by michoecho** [comment]
I'm 99% sure that server-side throughput is not a problem here, but that the concurrency is simply not high enough to sustain the throughput under the increased latency caused by repair. What was the concurrency (total number of cassandra-stress threads) and mean latency (as reported by cassandra-stress) in the test?

**Comment by roydahan** [comment]
From the above I can only answer the number of threads which is 50 threads * 4 loaders.
@fgelcer please provide the extra information requested above.

**Comment by michoecho** [comment]
To sustain 100k/s with 200 threads, the mean latency (client-visible service time) must be lower than 2.0 ms, so that's quite easy to exceed. The metrics say that the mean _server-side- write latency (which only contains the `storage_proxy` part — but the request must be picked up and answered too) is 1.5ms. If you add a single task quota (0.75ms) to this, you get 2.25ms.

The achieved throughput is 85k/s. If my theory is correct, this corresponds to a mean latency of 200 / (85k/s) = 2.35 ms. So it quite checks out.

In other words, with only 200 threads, you are testing if the mean latency is smaller than 2ms, not if the throughput can be sustained (if I'm right).

**Comment by fgelcer** [comment]
> Please post a monitor snapshot for me to download.

https://snapshots.raintank.io/dashboard/snapshot/5EW62aDBPXKJeF6O4gmvxGR86ozD24aN

**Comment by fgelcer** [comment]
> From the above I can only answer the number of threads which is 50 threads * 4 loaders. @fgelcer please provide the extra information requested above.

the exact c-s command for the write workload (that is the monitor attached to the issue description) is:
```
stress_cmd_w: "cassandra-stress write no-warmup cl=QUORUM duration=800m -schema 'replication(factor=3)' -mode cql3 native -rate 'threads=50 throttle=25000/s' -col 'size=FIXED(128) n=FIXED(8)' -pop 'dist=gauss(1..650000000,325000000,1200000)' "
```

so it does run 50 threads * 4 loaders
i will re-run the workload with 4 times more threads (so we will have 200 threads * 4 loaders). @michoecho , would that be enough? or should we go much bigger?

**Comment by fgelcer** [comment]
[running here](https://jenkins.scylladb.com/job/scylla-staging/job/fabio/job/perf-acceptance-criteria/job/perf-regression-latency-250gb-without-nemesis/5/) with 200 threads per loader (4 loaders)

**Comment by fgelcer** [comment]
i will be tracking it closely in the next hours, and will post a screenshot once we get to the stage where we had that drop previously

**Comment by fgelcer** [comment]
> [running here](https://jenkins.scylladb.com/job/scylla-staging/job/fabio/job/perf-acceptance-criteria/job/perf-regression-latency-250gb-with-nemesis/16/) with 200 threads per loader (4 loaders)

so, it failed with the same error message as described in https://github.com/scylladb/scylla-api-client/issues/31

all 3 workloads failed exact same way, and i'm using an AMI that i have already used in previous runs, and i'm using the same SCT `HEAD` as i had before (with my commits on top of it) without rebasing, so i don't understand why i'm getting this...
maybe @bhalevy knows?

**Comment by michoecho** [comment]
> so, it failed

The error says that a `latency_results.json` JSON file is missing or empty. A look into the SCT code says that this file is generated by `latency_calculator_decorator`, which is run on certain nemeses (the "ops"). The logs in the failed run say that it was run with `NoOpMonkey`, so presumably no ops happened, so no latency results file was generated?

(The code which generates and reads `latency_results` was written by @fgelcer, maybe he knows?  :stuck_out_tongue:)

> with the same error message as described in https://github.com/scylladb/scylla-api-client/issues/31

The error message only says that a JSON file is empty, the two issues are completely unrelated.

**Comment by fgelcer** [comment]
> > so, it failed
> 
> The error says that a `latency_results.json` JSON file is missing or empty. A look into the SCT code says that this file is generated by `latency_calculator_decorator`, which is run on certain nemeses (the "ops"). The logs in the failed run say that it was run with `NoOpMonkey`, so presumably no ops happened, so no latency results file was generated?
> 
> (The code which generates and reads `latency_results` was written by @fgelcer, maybe he knows? stuck_out_tongue)
> 
> > with the same error message as described in [scylladb/scylla-api-client#31](https://github.com/scylladb/scylla-api-client/issues/31)
> 
> The error message only says that a JSON file is empty, the two issues are completely unrelated.

well, that was actually a fault to be running with `NoOpMonkey`, but the test failed before the expected runtime for the c-s command.

i have a new one running (this time with the correct nemesis set) [in here](https://jenkins.scylladb.com/job/scylla-staging/job/fabio/job/perf-acceptance-criteria/job/perf-regression-latency-250gb-with-nemesis/17/)
